File: triage_cli/providers/openai.py
"""OpenAI Responses API provider — uses /responses endpoint with gpt-5.5."""
from __future__ import annotations

import json
import logging
from typing import Any

import httpx

from triage_cli.providers.unleash import (
    LLM_TIMEOUT_SECONDS,
    _base_url,
    _provider_error_message,
    _required_env,
)

logger = logging.getLogger(__name__)

DEFAULT_OPENAI_BASE_URL = "https://api.openai.com/v1"


class OpenAIResponsesProvider:
    name = "openai"

    async def complete(self, *, prompt: str, system_prompt: str, model: str) -> str:
        """Call the OpenAI Responses API and return output_text content."""
        endpoint = f"{_base_url('OPENAI_BASE_URL', DEFAULT_OPENAI_BASE_URL)}/responses"
        headers = {
            "Authorization": f"Bearer {_required_env('OPENAI_API_KEY', provider=self.name)}",
            "Accept": "application/json",
            "Content-Type": "application/json",
        }
        payload = {
            "model": model,
            "instructions": system_prompt,
            "input": prompt,
            "store": False,
        }

        data, response_headers = await _post_json(endpoint, headers=headers, payload=payload)
        text = _openai_text_from_response(data)
        if not text:
            request_id = _openai_request_id(data) or response_headers.get("x-request-id")
            suffix = f" RequestId: {request_id}." if request_id else ""
            raise RuntimeError(
                "OpenAI Responses API response did not include output_text." + suffix
            )
        return text


def _openai_text_from_response(data: Any) -> str:
    """Extract output_text from an OpenAI Responses API payload."""
    if not isinstance(data, dict):
        return ""
    top_level = data.get("output_text")
    if isinstance(top_level, str):
        return top_level

    output = data.get("output")
    if not isinstance(output, list):
        return ""

    chunks: list[str] = []
    for item in output:
        if not isinstance(item, dict):
            continue
        content = item.get("content")
        if not isinstance(content, list):
            continue
        for part in content:
            if not isinstance(part, dict) or part.get("type") != "output_text":
                continue
            text = part.get("text")
            if isinstance(text, str):
                chunks.append(text)
    return "".join(chunks)


def _openai_request_id(data: Any) -> str | None:
    if isinstance(data, dict):
        value = data.get("id")
        return value if isinstance(value, str) and value else None
    return None


async def _post_json(
    endpoint: str,
    *,
    headers: dict[str, str],
    payload: dict[str, Any],
) -> tuple[Any, dict[str, str]]:
    try:
        async with httpx.AsyncClient(timeout=LLM_TIMEOUT_SECONDS) as client:
            response = await client.post(endpoint, headers=headers, json=payload)
    except httpx.RequestError as e:
        raise RuntimeError(f"LLM provider API call failed: {e}") from e

    if response.status_code >= 400:
        raise RuntimeError(_provider_error_message(response))

    try:
        data = response.json()
    except json.JSONDecodeError as e:
        raise RuntimeError("LLM provider API response was not valid JSON.") from e
    return data, dict(response.headers)

File: triage-cli-rs/src/providers/openai.rs
//! OpenAI Responses API provider.
//!
//! Hits the `/responses` endpoint, treating `system_prompt` as
//! `instructions` and `prompt` as `input`. Behavior tracks the Python
//! implementation; output text is extracted from `output_text` (or the
//! nested `output[].content[].text` array if the top-level shortcut is
//! absent).

use std::future::Future;
use std::pin::Pin;
use std::time::Duration;

use reqwest::Client;
use serde_json::{json, Value};

use super::{base_url, required_env, LlmProvider, ProviderError};
use super::unleash::{provider_error_message, LLM_TIMEOUT_SECS};

const DEFAULT_BASE_URL: &str = "https://api.openai.com/v1";

pub struct OpenAiProvider;

impl LlmProvider for OpenAiProvider {
    fn name(&self) -> &'static str {
        "openai"
    }

    fn complete<'a>(
        &'a self,
        prompt: &'a str,
        system_prompt: &'a str,
        model: &'a str,
    ) -> Pin<Box<dyn Future<Output = Result<String, ProviderError>> + Send + 'a>> {
        Box::pin(async move {
            let endpoint = format!("{}/responses", base_url("OPENAI_BASE_URL", DEFAULT_BASE_URL));
            let api_key = required_env("OPENAI_API_KEY", "openai")?;
            let payload = json!({
                "model": model,
                "instructions": system_prompt,
                "input": prompt,
                "store": false,
            });

            let client = Client::builder()
                .timeout(Duration::from_secs(LLM_TIMEOUT_SECS))
                .build()
                .map_err(|e| ProviderError::Transport(e.to_string()))?;
            let resp = client
                .post(&endpoint)
                .header("Authorization", format!("Bearer {api_key}"))
                .header("Accept", "application/json")
                .header("Content-Type", "application/json")
                .json(&payload)
                .send()
                .await
                .map_err(|e| ProviderError::Transport(e.to_string()))?;
            let status = resp.status();
            let response_headers = resp.headers().clone();
            if status.as_u16() >= 400 {
                let rid = response_headers
                    .get("x-request-id")
                    .and_then(|v| v.to_str().ok())
                    .map(|s| s.to_string());
                let body = resp.text().await.unwrap_or_default();
                return Err(provider_error_message(status.as_u16(), &body, rid.as_deref()));
            }
            let body = resp
                .text()
                .await
                .map_err(|e| ProviderError::Transport(e.to_string()))?;
            let data: Value = serde_json::from_str(&body).map_err(|_| ProviderError::NonJson)?;
            let text = openai_text_from_response(&data);
            if text.is_empty() {
                let rid = openai_request_id(&data)
                    .or_else(|| {
                        response_headers
                            .get("x-request-id")
                            .and_then(|v| v.to_str().ok())
                            .map(|s| s.to_string())
                    })
                    .map(|r| format!(" RequestId: {r}."))
                    .unwrap_or_default();
                return Err(ProviderError::NoText("OpenAI Responses", rid));
            }
            Ok(text)
        })
    }
}

fn openai_text_from_response(data: &Value) -> String {
    if let Some(top) = data.get("output_text").and_then(Value::as_str) {
        return top.to_string();
    }
    let Some(output) = data.get("output").and_then(Value::as_array) else {
        return String::new();
    };
    let mut chunks: Vec<String> = Vec::new();
    for item in output {
        let Some(content) = item.get("content").and_then(Value::as_array) else {
            continue;
        };
        for part in content {
            if part.get("type").and_then(Value::as_str) == Some("output_text") {
                if let Some(s) = part.get("text").and_then(Value::as_str) {
                    chunks.push(s.to_string());
                }
            }
        }
    }
    chunks.join("")
}

fn openai_request_id(data: &Value) -> Option<String> {
    data.get("id")
        .and_then(Value::as_str)
        .filter(|s| !s.is_empty())
        .map(|s| s.to_string())
}
